csharp/rust/src/conreq.rs

```diff
@@ -0,0 +1,240 @@
+        match self {
+            EProtocolVersion::None => None,
+            EProtocolVersion::RESP2 => Some(redis::ProtocolVersion::RESP2),
+            EProtocolVersion::RESP3 => Some(redis::ProtocolVersion::RESP3),
+        }
+    }
+}
+
+#[repr(C)]
+pub enum ETlsMode {
+    None = 0,
+    NoTls = 1,
+    InsecureTls = 2,
+    SecureTls = 3,
+}
+
+impl ETlsMode {
+    pub fn to_redis(self) -> Option<client::TlsMode> {
+        match self {
+            ETlsMode::None => None,
+            ETlsMode::NoTls => Some(client::TlsMode::NoTls),
+            ETlsMode::InsecureTls => Some(client::TlsMode::InsecureTls),
+            ETlsMode::SecureTls => Some(client::TlsMode::SecureTls),
+        }
+    }
+}
+
+#[repr(C)]
+pub struct ConnectionRetryStrategy {
+    pub ignore: c_int,
+    pub exponent_base: c_uint,
+    pub factor: c_uint,
+    pub number_of_retries: c_uint,
+}
+
+impl ConnectionRetryStrategy {
+    pub fn to_redis(self) -> Option<client::ConnectionRetryStrategy> {
+        if self.ignore == 0 {
+            return Some(client::ConnectionRetryStrategy {
+                exponent_base: self.exponent_base,
+                factor: self.factor,
+                number_of_retries: self.number_of_retries,
+            });
+        } else {
+            None
+        }
+    }
+}
+
+#[repr(C)]
+pub struct PeriodicCheck {
+    pub kind: EPeriodicCheckKind,
+    pub secs: c_ulonglong,
+    pub nanos: c_ulonglong,
+}
+#[repr(C)]
+pub enum EPeriodicCheckKind {
+    None = 0,
+    Enabled = 1,
+    Disabled = 2,
+
+    /// secs and nanos on PeriodicCheck must be set
+    ManualInterval = 3,
+}
+
+impl PeriodicCheck {
+    pub fn to_redis(self) -> Option<client::PeriodicCheck> {
+        match self.kind {
+            EPeriodicCheckKind::None => None,
+            EPeriodicCheckKind::Enabled => Some(client::PeriodicCheck::Enabled),
+            EPeriodicCheckKind::Disabled => Some(client::PeriodicCheck::Disabled),
+            EPeriodicCheckKind::ManualInterval => Some(client::PeriodicCheck::ManualInterval(
+                Duration::from_secs(self.secs).add(Duration::from_nanos(self.nanos)),
+            )),
+        }
+    }
+}
+
+#[repr(C)]
+pub struct OptionalU32 {
+    pub ignore: c_int,
+    pub value: c_uint,
+}
+impl Into<Option<u32>> for OptionalU32 {
+    fn into(self) -> Option<u32> {
+        if self.ignore == 0 {
+            Some(self.value)
+        } else {
+            None
+        }
+    }
+}
+#[repr(C)]
+pub struct OptionalU64 {
+    pub ignore: c_int,
+    pub value: c_ulonglong,
+}
+
+impl Into<Option<u64>> for OptionalU64 {
+    fn into(self) -> Option<u64> {
+        if self.ignore == 0 {
+            Some(self.value)
+        } else {
+            None
+        }
+    }
+}
+
+#[repr(C)]
+pub struct ConnectionRequest {
+    pub read_from: ReadFrom,
+    pub client_name: *const c_char,
+    pub auth_username: *const c_char,
+    pub auth_password: *const c_char,
+    pub database_id: c_longlong,
```

Comment:
optional:
```suggestion
    pub database_id: c_uint,
```
There are 64 DBs only or so

Reply:
in c, `unsigned int` is not guaranteed to be 64bit, which is why i decided to use `long long` here. Feel free to call on a decision.